File: internal/store/marks.go
package store

import (
	"database/sql"
	"fmt"
)

// The two hand-set company marks, orthogonal to the verdict:
//   - flagged_at  — a bookmark, toggled on/off
//   - reviewed_at — a last-reviewed stamp, refreshed on every "Mark reviewed"
//
// Both return sql.ErrNoRows for an unknown company so handlers can answer 404
// rather than silently no-op.

// SetFlagged flags (flagged_at = now) or unflags (flagged_at = NULL) a company.
func (db *DB) SetFlagged(companyID string, flagged bool) error {
	q := `UPDATE companies SET flagged_at = NULL WHERE id = ?`
	if flagged {
		q = `UPDATE companies SET flagged_at = CURRENT_TIMESTAMP WHERE id = ?`
	}
	return db.execMark(q, companyID, "flagged")
}

// TouchReviewed stamps a company as reviewed now. Repeated calls keep moving
// the stamp forward — the table sorts on it, so the user can cycle through
// companies oldest-reviewed-first instead of always landing on the same ones.
func (db *DB) TouchReviewed(companyID string) error {
	return db.execMark(`UPDATE companies SET reviewed_at = CURRENT_TIMESTAMP WHERE id = ?`, companyID, "reviewed")
}

func (db *DB) execMark(q, companyID, what string) error {
	res, err := db.Exec(q, companyID)
	if err != nil {
		return fmt.Errorf("set %s %s: %w", what, companyID, err)
	}
	n, err := res.RowsAffected()
	if err != nil {
		return fmt.Errorf("set %s %s: %w", what, companyID, err)
	}
	if n == 0 {
		return sql.ErrNoRows
	}
	return nil
}

File: scout/store/marks.py
"""Hand-set company marks (flagged / reviewed). Port of internal/store/marks.go.

Both raise NotFound for an unknown company so handlers can answer 404 rather
than silently no-op.
"""
from __future__ import annotations

import sqlite3

from . import errors


def set_flagged(con: sqlite3.Connection, company_id: str, flagged: bool) -> None:
    """Flag (flagged_at = now) or unflag (flagged_at = NULL) a company."""
    if flagged:
        q = "UPDATE companies SET flagged_at = CURRENT_TIMESTAMP WHERE id = ?"
    else:
        q = "UPDATE companies SET flagged_at = NULL WHERE id = ?"
    _exec_mark(con, q, company_id)


def touch_reviewed(con: sqlite3.Connection, company_id: str) -> None:
    """Stamp a company as reviewed now (repeated calls move the stamp forward)."""
    _exec_mark(con, "UPDATE companies SET reviewed_at = CURRENT_TIMESTAMP WHERE id = ?", company_id)


def _exec_mark(con: sqlite3.Connection, q: str, company_id: str) -> None:
    cur = con.execute(q, (company_id,))
    if cur.rowcount == 0:
        raise errors.NotFound()
